lightning/src/util/test_utils.rs

```diff
@@ -269,6 +271,108 @@ impl<'a> chain::Watch<EnforcingSigner> for TestChainMonitor<'a> {
+	/// amount, and commitment number so we can build the justice tx after our counterparty
+	/// revokes it.
+	unsigned_justice_tx_data: Mutex<HashMap<OutPoint, VecDeque<JusticeTxData>>>,
+	/// After receiving a revoke_and_ack for a commitment number, we'll form and store the justice
+	/// tx which would be used to provide a watchtower with the data it needs.
+	watchtower_state: Mutex<HashMap<OutPoint, HashMap<Txid, Transaction>>>,
+	destination_script: Script,
+}
+
+impl WatchtowerPersister {
+	pub(crate) fn new(destination_script: Script) -> Self {
+		WatchtowerPersister {
+			persister: TestPersister::new(),
+			unsigned_justice_tx_data: Mutex::new(HashMap::new()),
+			watchtower_state: Mutex::new(HashMap::new()),
+			destination_script,
+		}
+	}
+
+	pub(crate) fn justice_tx(&self, funding_txo: OutPoint, commitment_txid: &Txid)
+	-> Option<Transaction> {
+		self.watchtower_state.lock().unwrap().get(&funding_txo).unwrap().get(commitment_txid).cloned()
+	}
+
+	fn form_justice_data_from_commitment(&self, counterparty_commitment_tx: &CommitmentTransaction)
+	-> Option<JusticeTxData> {
+		let output_idx = counterparty_commitment_tx.revokeable_output_index()?;
+		let trusted_tx = counterparty_commitment_tx.trust();
+		let built_tx = trusted_tx.built_transaction();
+		let value = built_tx.transaction.output[output_idx as usize].value;
+		let justice_tx = counterparty_commitment_tx.build_to_local_justice_tx(
+			FEERATE_FLOOR_SATS_PER_KW as u64, self.destination_script.clone()).ok()?;
+		let commitment_number = counterparty_commitment_tx.commitment_number();
+		Some(JusticeTxData { justice_tx, value, commitment_number })
+	}
+}
+
+impl<Signer: sign::WriteableEcdsaChannelSigner> chainmonitor::Persist<Signer> for WatchtowerPersister {
+	fn persist_new_channel(&self, funding_txo: OutPoint,
+		data: &channelmonitor::ChannelMonitor<Signer>, id: MonitorUpdateId
+	) -> chain::ChannelMonitorUpdateStatus {
+		let res = self.persister.persist_new_channel(funding_txo, data, id);
+
+		assert!(self.unsigned_justice_tx_data.lock().unwrap()
+			.insert(funding_txo, VecDeque::new()).is_none());
+		assert!(self.watchtower_state.lock().unwrap()
+			.insert(funding_txo, HashMap::new()).is_none());
+
+		let initial_counterparty_commitment_tx = data.initial_counterparty_commitment_tx()
+			.expect("First and only call expects Some");
+		if let Some(justice_data)
+			= self.form_justice_data_from_commitment(&initial_counterparty_commitment_tx) {
+			self.unsigned_justice_tx_data.lock().unwrap()
+				.get_mut(&funding_txo).unwrap()
+				.push_back(justice_data);
+		}
+		res
+	}
+
+	fn update_persisted_channel(
+		&self, funding_txo: OutPoint, update: Option<&channelmonitor::ChannelMonitorUpdate>,
+		data: &channelmonitor::ChannelMonitor<Signer>, update_id: MonitorUpdateId
+	) -> chain::ChannelMonitorUpdateStatus {
+		let res = self.persister.update_persisted_channel(funding_txo, update, data, update_id);
+
+		if let Some(update) = update {
```

Comment:
Sure, will try and get to this over the next couple of days

Reply:
Sorry for the delay here, getting to this now. Were you thinking to expose basically a cleaned up version of this test util? I'm wondering who would this be for, since I imagine someone using this interface would want to return a different `ChannelMonitorUpdateStatus` depending on however they're using these signed justice txs (e.g. sending to remote watchtower/other server).

Hmm although... I could extend this util to take some user-implemented interface that takes a signed justice tx, does something, and returns a `ChannelMonitorUpdateStatus`, so our util handles the tracking and the user can focus on handling what to do with the tx? Would that make sense?

Reply:
Ah ok that makes sense!